src/Build/Evaluation/LazyItemEvaluator.RemoveOperation.cs

```diff
@@ -23,9 +23,7 @@ public RemoveOperation(RemoveOperationBuilder builder, LazyItemEvaluator<P, I, M
                 _matchOnMetadataOptions = builder.MatchOnMetadataOptions;
             }
 
-            // todo port the self referencing matching optimization (e.g. <I Remove="@(I)">) from Update to Remove as well. Ideally make one mechanism for both. https://github.com/Microsoft/msbuild/issues/2314
-            // todo Perf: do not match against the globs: https://github.com/Microsoft/msbuild/issues/2329
-            protected override ImmutableList<I> SelectItems(ImmutableList<ItemData>.Builder listBuilder, ImmutableHashSet<string> globsToIgnore)
+            protected override void ApplyImpl(ImmutableList<ItemData>.Builder listBuilder, ImmutableHashSet<string> globsToIgnore)
```

Comment:
"For future generations" nit: ApplyImpl isn't exactly a self-documenting function name (like `ItemspecContainsASingleItemReference`), so I'd suggest writing up a short summary describing its purpose. Or adding that summary to the base function while you have the info in your head.